### ISSUE TITLE
Glide scans wrong version of dependency to find transitive dependencies

### ISSUE THREAD

**Issue by seh** [Masterminds/glide#624]
It appears that when _glide update_ scans a cached dependency to find its transitive dependencies, at least for Git repositories it scans the source as it appears in the "master" branch, even when a _glide.yaml_ file specifies a different "revision" for that repository.

I attempted to work around go-kit/kit#366 by going into the Git repository cloned into the _~/.glide/cache/src/https-github.com-go-kit-kit_ directory, checking out [the "0.2.0" tag](https://github.com/go-kit/kit/tree/v0.2.0), creating a new branch there, editing a file to comment out some problematic code, committing the change to my new branch, and editing _glide.yaml_ to nominate my new branch there as the "revision" to use.

If I leave my Git repository in the Glide cache sitting with my new branch checked out, [_glide update_ fails](https://gitter.im/Masterminds/glide?at=57ee80d334a8d5681ccf102e).
```text
[WARN]    Download failed.
[ERROR]    Update failed for github.com/go-kit/kit: Unable to update repository
[ERROR]    Failed to do initial checkout of config: Unable to update repository
```
If I change that repository to check out branch "master" (_git checkout master_) and run _glide update_, the above problem is resolved, but then Glide reads the repository files on the "master" branch rather than my new, corrected branch.

Apparently Glide _does_ check out my corrected branch at some point, because when I visit the Git repository after _glide update_ fails, I find my corrected branch checked out. However, Glide still fails due to trying to fetch an import path that does not exist, even though my corrected file checked in to my new branch (again, the one specified as the "revision" in my _glide.yaml_ file) no longer contains any reference to that missing import path.

**Comment by seh** [comment]
For the record, [the same thing occurs](https://gitter.im/Masterminds/glide?at=57ee98ccd38f186520b457a7) when I don't touch the Git repository in the cache, but rather point at a local Git repository outside the cache via a URL with the "file" scheme.

**Comment by seh** [comment]
I'm still learning my way around the Glide code base, but I'll record some findings. In [file _repo/installer.go_'s method `VersionHandler.SetVersion`](https://github.com/Masterminds/glide/blob/3844a21c7453dd09d02f87c5b3f6e8ab9717b2be/repo/installer.go#L719), we see that it tells the VCS to switch to the intended version near the end with [the call to `VcsVersion`](https://github.com/Masterminds/glide/blob/3844a21c7453dd09d02f87c5b3f6e8ab9717b2be/repo/installer.go#L783).

Over in file _dependency/resolver.go_, we see that [we call on `VersionHandler.Process`](https://github.com/Masterminds/glide/blob/bd55e763dbfac7b4684f6f3149bb56bed313175a/dependency/resolver.go#L491) before [we call on `VersionHandler.SetVersion`](https://github.com/Masterminds/glide/blob/bd55e763dbfac7b4684f6f3149bb56bed313175a/dependency/resolver.go#L533). However, it looks like that sequence arises only when the package didn't exist, and [we fetch it in between](https://github.com/Masterminds/glide/blob/bd55e763dbfac7b4684f6f3149bb56bed313175a/dependency/resolver.go#L525) the aforementioned `Process` and `SetVersion` calls. The dependency winds up back in the queue to enjoy a subsequent call to `Process` again, this time _after_ the preceding call to `SetVersion`.

Given that reading, we shouldn't wind up processing a repository before having set it to the intended version.

In my example, we can see Glide checking out the Git repository and changing branches, but for some reason the import path scanning winds up reading content only available on the "master" branch.
```shell
% pushd ~/.glide/cache/src/file--Users-seh-go-src-github.com-go-kit-kit 
~/.glide/cache/src/file--Users-seh-go-src-github.com-go-kit-kit ~/go/src/github.com/go-kit/kit

% git reflog
2dcf581 HEAD@{0}: checkout: moving from master to seh-fix
df0518c HEAD@{1}: clone: from file:///Users/seh/go/src/github.com/go-kit/kit

% git branch
  master
* seh-fix
```

**Comment by seh** [comment]
Forget it. This was my fault.

I had a previous _vendor_ directory that I had moved "out of the way", having renamed it to _vendor-good_. Glide was scanning the files in that directory, and _that's_ where it was finding the offending file in _go-kit_'s "master" branch version—not within Glide's cache. Once I moved this _vendor-good_ directory outside of my project directory tree, _glide install_ completed as expected.

Sorry for the distraction.
